### ISSUE TITLE
`:defer` problems with exception reporting and propogation

### ISSUE THREAD

**Issue by errael** [vim/vim#13382]
### Steps to reproduce

(This is copied from #13376)

I've been curious as to why, in 
`Errors are reported but do not cause aborting execution of deferred functions.`
the word "reported" is used.

I ask because **two exceptions are thrown**, but only **one is reported**. The first exception is lost. Run the script below; to get simpler output, get rid of the `try/catch/finally` and just run Foo().

Note: `&verbose = 14`
```
Throwing TestException
Exception thrown: TestException
Exception thrown: InnerException
Exception made pending: InnerException
finally [30, 31, 20, 9990, 10, 11]
v:exception:
Exception resumed: InnerException
Exception discarded: InnerException
Error detected while processing :source buffer=1[38]..:source buffer=1[32]..function <SNR>11_Foo[5].
.<SNR>11_DeferWithExcept[2]..<SNR>11_Bar:
line    2:
E605: Exception not caught: InnerException
```

```
vim9script

&verbose = 14

var callTrace: list<number> = []
def Bar()
  callTrace += [9990]
  throw 'InnerException'
enddef

def Defer(tag: number)
  callTrace += [tag]
  callTrace += [tag + 1]
enddef

def DeferWithExcept(tag: number)
  callTrace += [tag]
  Bar()
  callTrace += [tag + 1]
enddef


def Foo()
  defer Defer(10)
  defer DeferWithExcept(20)
  defer Defer(30)
  echo "Throwing TestException"
  throw "TestException"
enddef

try
  Foo()
catch /TestException/
  callTrace += [7]
finally
  echo "finally" callTrace
  echo "v:exception:" v:exception
endtry

echo "finished" callTrace
```

### Expected behaviour

(more like suggested behavior, I just don't know)

Exceptions in a deferred function do not propagate beyond the function.
They do not interfere with the regular flow of exception handling.

It as though they run like
```
try
    DeferredFunction()
catch
    report_exception()
endtry
```

It's kind of like the deferred functions are run as separate threads, an exception in one thread does not affect the execution of other threads. That seems to fit with the idea that if the current deferred function is aborted, then the other deferred functions are executed.

I just read (some of) the `:help exception-handling`, wow.

### Version of Vim

9.0.2050 (but `:version` shows 2049)

### Environment

ubuntu

**Comment by yegappan** [comment]
I have created a draft PR #13386 to address this problem.  Can you try that out?

**Comment by errael** [comment]
Looks pretty good. See following post for examples.

One thing that might be useful is an indication of which exceptions are
coming from deferred functions. The name of the function is already in
the exception trace. If it's worthwhile, here are two possibilities as examples

1. Adjust the existing message.
    ```
    Error detected while processing deferred function ...
    ```
2. Precede the message with something on it's own line
    ```
    Deferred function error.
    Error detected while processing ...
    ```

The trace for deferred functions doesn't have the line where the defer was posted; it seems like if it's triggered by an exception you see the exception line, when triggered by function exit you see the first line of the function. Don't know if it's worth adding this information, but it's something to notice. Hmm, might argue for using (2.) and then "deferred from" information could be more easily displayed.

**Comment by errael** [comment]
Here's several examples based on the following script, using #13386.
```
vim9script

var callTrace: list<number> = []
def Except(exc_tag: number)
  callTrace += [exc_tag]
  throw 'InnerException'
enddef

def FirstDefer(tag: number)
  callTrace += [tag]
  callTrace += [tag + 1]
  echo "FirstDefer" callTrace
enddef

def Defer(tag: number)
  callTrace += [tag]
  callTrace += [tag + 1]
enddef

def DeferWithExcept1(tag: number, exc_tag = 9999)
  callTrace += [tag] | Except(exc_tag) | callTrace += [tag + 1]
enddef

def DeferWithExcept2(tag: number, exc_tag = 9999)
  callTrace += [tag] | Except(exc_tag) | callTrace += [tag + 1]
enddef

def DeferWithExcept3(tag: number, exc_tag = 9999)
  callTrace += [tag] | Except(exc_tag) | callTrace += [tag + 1]
enddef

def Bar()
  defer Defer(110)
  defer DeferWithExcept2(120, 9995)
  defer Defer(130)
  echo "Throwing BarException"
  throw "BarException"
enddef

def Foo()
  defer FirstDefer(10)
  defer DeferWithExcept2(20, 9991)
  defer Defer(30)
  echo "Throwing TestException"
  throw "TestException"
enddef

#Foo()

try
  Foo()
catch /TestException/
  callTrace += [7]
finally
  echo "   finally" callTrace
endtry

echo "  finished" callTrace
```

---
Output from the unmodified script. The triggered exception is caught.

```
Throwing TestException
Error detected while processing :source buffer=1[54]..function <SNR>11_Foo[8]..:source buffer=1[54].
.function <SNR>11_Foo[8]..<SNR>11_DeferWithExcept2[1]..<SNR>11_Except:
line    2:
E605: Exception not caught: InnerException
FirstDefer [30, 31, 20, 9991, 10, 11]
   finally [30, 31, 20, 9991, 10, 11, 7]
  finished [30, 31, 20, 9991, 10, 11, 7]
```

--- 
Output from running `Foo()` directly, no `try`/`catch`.
```
Throwing TestException
Error detected while processing :source buffer=1[51]..function <SNR>11_Foo[8]..:source buffer=1[51].
.function <SNR>11_Foo[8]..<SNR>11_DeferWithExcept2[1]..<SNR>11_Except:
line    2:
E605: Exception not caught: InnerException
FirstDefer [30, 31, 20, 9991, 10, 11]
Error detected while processing :source buffer=1[51]..:source buffer=1[51]..function <SNR>11_Foo:
line    8:
E605: Exception not caught: TestException
```

--- 
Output from taking out the throw from `Foo()`, normal function exit
```
def Foo()
  defer FirstDefer(10)
  defer DeferWithExcept2(20, 9991)
  defer Defer(30)
enddef
```
POSSIBLE PROBLEM: exception from deferred function leaks into main processing.
```
Error detected while processing :source buffer=1[51]..function <SNR>11_Foo[1]..:source buffer=1[51].
.function <SNR>11_Foo[1]..<SNR>11_DeferWithExcept2[1]..<SNR>11_Except:
line    2:
E605: Exception not caught: InnerException
FirstDefer [30, 31, 20, 9991, 10, 11]
```

---
Output when deferring the same function three times. Change `Foo()` to
```
def Foo()
  defer DeferWithExcept2(10, 9991)
  defer DeferWithExcept2(20, 9992)
  defer DeferWithExcept2(30, 9993)
  echo "Throwing TestException"
  throw "TestException"
enddef
```
Interesting/nice how it doesn't repeat the same trace.
Note that if the trace had the line number, it might not be able to optimize
out duplicate traces.
```
Throwing TestException
Error detected while processing :source buffer=1[54]..function <SNR>11_Foo[8]..:source buffer=1[54].
.function <SNR>11_Foo[8]..<SNR>11_DeferWithExcept2[1]..<SNR>11_Except:
line    2:
E605: Exception not caught: InnerException
E605: Exception not caught: InnerException
E605: Exception not caught: InnerException
   finally [30, 9993, 20, 9992, 10, 9991, 7]
  finished [30, 9993, 20, 9992, 10, 9991, 7]
```

---
Ouput when `Foo()` directly calls function `Bar()` and `Bar()` `:defer`s some functions;
`Bar()` throws an excecption as it's last line, which causes `Foo()` to abort.

```
def Foo()
  defer FirstDefer(10)
  defer DeferWithExcept2(20, 9991)
  defer Defer(30)
  Bar()
  echo "Throwing TestException"
  throw "TestException"
enddef
```

```
Throwing BarException
Error detected while processing :source buffer=1[52]..function <SNR>11_Foo[4]..<SNR>11_Bar[5]..:sour
ce buffer=1[52]..function <SNR>11_Foo[4]..<SNR>11_Bar[5]..<SNR>11_DeferWithExcept2[1]..<SNR>11_Excep
t:
line    2:
E605: Exception not caught: InnerException
Error detected while processing :source buffer=1[52]..function <SNR>11_Foo[4]..:source buffer=1[52].
.function <SNR>11_Foo[4]..<SNR>11_DeferWithExcept2[1]..<SNR>11_Except:
line    2:
E605: Exception not caught: InnerException
FirstDefer [130, 131, 120, 9995, 110, 111, 30, 31, 20, 9991, 10, 11]
Error detected while processing :source buffer=1[52]..:source buffer=1[52]..function <SNR>11_Foo[4].
.<SNR>11_Bar:
line    5:
E605: Exception not caught: BarException
```

**Comment by chrisbra** [comment]
should be fixed by 0ab500dede4edd8d5aee7ddc63444537be527871
